### ISSUE TITLE
Set up a data center closer to Iran

### ISSUE THREAD

**Issue by myleshorton** [getlantern/lantern#2210]

**Comment by myleshorton** [comment]
We could try the Vultr Frankfurt data center:

https://www.vultr.com/locations/

**Comment by myleshorton** [comment]
Or India through these guys:

http://vps.net/cloud-datacenter-locations/asian-datacenters

This site is really useful:

http://serverbear.com/

**Comment by oxtoacart** [comment]
@myleshorton Nice finds!

**Comment by myleshorton** [comment]
Just wanted to note that the key is really setting up our data center **close to CloudFlare's data center** that gets traffic from Iran. We should test, but could be Milan or Vienna, for example.

**Comment by fffw** [comment]
A global ping test like below might be helpful to determine which data center has lowest delay, and to which cloudfront node.

>  http://www.super-ping.com/?ping=roundrobin.getiantem.org&locale=en

>  http://startping.com/ping/roundrobin.getiantem.org/54d3f1b68c3dcfe80afcb0e1

**Comment by myleshorton** [comment]
Great idea @fffw!

On Thu, Feb 5, 2015 at 2:58 PM, fffw <notifications@github.com> wrote:

> A global ping test like below might be helpful to determine which data
> center has lowest delay, and to which cloudfront node.
>
> http://www.super-ping.com/?ping=roundrobin.getiantem.org&locale=en
>
> http://startping.com/ping/roundrobin.getiantem.org/54d3f1b68c3dcfe80afcb0e1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2210#issuecomment-73148372>.
>



-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by oxtoacart** [comment]
Neat idea!  Few things to consider:

Clients don't connect to roundrobin.getiantem.org, they connect to some
masquerade domains using hardcoded IPs in masquerades.go.  It would be
interesting to run this test for at least a sampling of those to see what
the latencies to CloudFlare are like.

As Adam mentioned, the other thing to consider is the latency from
CloudFlare to our VPSs.  I'm not sure of the best way to determine that,
since geolocating CloudFlare is a little tricky due to the use of anycast.

Cheers,
Ox

-----------------------------------------------------------------------------------------

"I love people who harness themselves, an ox to a heavy cart,
who pull like water buffalo, with massive patience,
who strain in the mud and the muck to move things forward,
who do what has to be done, again and again."

- Marge Piercy

**Comment by fffw** [comment]
Neat idea!  Few things to consider:


Clients don't connect to roundrobin.getiantem.org, they connect to some

masquerade domains using hardcoded IPs in masquerades.go.  It would be

interesting to run this test for at least a sampling of those to see what

the latencies to CloudFlare are like.



We can automate if we want, might be quicker than manual operation on samples.


As Adam mentioned, the other thing to consider is the latency from

CloudFlare to our VPSs.  I'm not sure of the best way to determine that,

since geolocating CloudFlare is a little tricky due to the use of anycast.




In fact, my original idea is to test latency between CloudFlare and potential VPS, but in reverse path. Those probes are always located at some data center.




It might be hard to have PCs in Iran to test the latency between client and CloudFlare. We may add it to statistics report later, if not already.



Cheers,

Ox


-----------------------------------------------------------------------------------------


"I love people who harness themselves, an ox to a heavy cart,

who pull like water buffalo, with massive patience,

who strain in the mud and the muck to move things forward,

who do what has to be done, again and again."


- Marge Piercy



On Thu, Feb 5, 2015 at 10:15 PM, myleshorton <notifications@github.com>

wrote:


> Great idea @fffw!

>

**Comment by myleshorton** [comment]
We could do a pilot with OTF Cloud on this

**Comment by oxtoacart** [comment]
From running some traceroutes, it looks like traffic from Iran to cloudflare.com tends to be routed to Europe (London, Frankfurt, Paris).

**Comment by myleshorton** [comment]
Yup that's what I suspected - interesting it goes to multiple though. Does
that depend on the domain?

On Thursday, February 26, 2015, oxtoacart <notifications@github.com> wrote:

> From running some traceroutes, it looks like traffic from Iran to
> cloudflare.com tends to be routed to Europe (London, Frankfurt, Paris).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2210#issuecomment-76203851>.
>


-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by myleshorton** [comment]
Sounds like the DO Amsterdam data center might not be so bad.

On Thursday, February 26, 2015, Adam Fisk <afisk@getlantern.org> wrote:

> Yup that's what I suspected - interesting it goes to multiple though. Does
> that depend on the domain?
>
> On Thursday, February 26, 2015, oxtoacart <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> From running some traceroutes, it looks like traffic from Iran to
>> cloudflare.com tends to be routed to Europe (London, Frankfurt, Paris).
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/getlantern/lantern/issues/2210#issuecomment-76203851>
>> .
>>
>
>
> --
> --
> President
> Brave New Software Project, Inc.
> https://www.getlantern.org
> pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89
>


-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by oxtoacart** [comment]
Yeah, I'm trying the Amsterdam datacenter as we speak.

Regarding the routing - no, I see different routes used all for the `cloudflare.com` domain.  It's possible that they're all ending up at the same CloudFlare data center and just taking different routes there, but for now I think just placing servers in Europe is close enough.

I'm running tests with boom from Iran which will tell us more shortly.

**Comment by atavism** [comment]
Are you on testing from our r0ute.com VPS? I was going to build flashlight on there and run some tests, but it looks like you already did (or started to).

**Comment by oxtoacart** [comment]
Yes, I'm testing from the r0ute.com VPS.

**Comment by atavism** [comment]
Do you want to pair on this ticket? I was going to work on this. Do you want me to setup some flashlight servers in the DO Amsterdam data center to start?

**Comment by myleshorton** [comment]
I'm also curious about the vultr Milan data center and just generally
testing different options.

On Thursday, February 26, 2015, atavism <notifications@github.com> wrote:

> Do you want to pair on this ticket? I was going to work on this. Do you
> want me to setup some flashlight servers in the DO Amsterdam data center to
> start?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2210#issuecomment-76215112>.
>


-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by oxtoacart** [comment]
@atavism Yeah.  If you want to modify lantern_aws so that we can target the Amsterdam 3 data center when we spool up new flashlight instances, I'll take a look at the peerscanner changes.  We've been using this naming convention for flashlight:

fl-<cc>-yyyymmdd-###

Where <cc> is the two digit country code where the servers are located, yyyymmdd is the date the server was launched and ### is a series number.  So we might have:

```
fl-sg-20150226-001
fl-sg-20150226-002
fl-sg-20150226-003
```

Or

```
fl-nl-20150226-001
fl-nl-20150226-002
fl-nl-20150226-003
```

I'll set up peerscanner to automatically create rotations based on the country code, e.g. `nl.fallbacks.getiantem.org` and `sg.fallbacks.getiantem.org`.

Sound good?

**Comment by oxtoacart** [comment]
@myleshorton I agree it would be interesting to explore other provides, but if you look at the results that I just emailed, the Amsterdam data center performs very well indeed from Iran.  Since we already have the ability to deploy to Digital Ocean, I would say we start with that and explore the other stuff if we have time.

**Comment by myleshorton** [comment]
Yeah that's fine. I'm more talking about when we really dive into the speed
iteration.

On Thursday, February 26, 2015, oxtoacart <notifications@github.com> wrote:

> @myleshorton <https://github.com/myleshorton> I agree it would be
> interesting to explore other provides, but if you look at the results that
> I just emailed, the Amsterdam data center performs very well indeed from
> Iran. Since we already have the ability to deploy to Digital Ocean, I would
> say we start with that and explore the other stuff if we have time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2210#issuecomment-76216235>.
>


-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by atavism** [comment]
> Sound good?

Yep. I'm looking into the lantern_aws changes now.

**Comment by oxtoacart** [comment]
@atavism Let me know if you have any questions about the lantern_aws stuff.  @aranhoide knows the most about it, but I've had my hands in there.  I think what you need to do is add a provider to salt/salt_cloud/cloud.providers that's like do_sg but is do_nl.

**Comment by oxtoacart** [comment]
Oh, you'll need to add a profile to cloud_profiles too.

**Comment by oxtoacart** [comment]
@atavism Actually, looks like the do provider takes a configuration parameter (grain) with the DO region.  That parameter appears to be set in bin/config.py and you can override it in your bin/config_overrides.py.  If we're lucky, that might be all that's necessary.

**Comment by atavism** [comment]
@oxtoacart Yep, changed all of those and created a new profile for Amsterdam. Waiting for SQS to process the requests I made and for the servers to show up in DO. Should be finished shortly.

**Comment by atavism** [comment]
Ah, ```config_overrides.py``` is used to override the options set in ```config.py```. Makes sense. It might be useful to update ```fake_controller.py``` to accept a profile parameter option too.

**Comment by oxtoacart** [comment]
@atavism I'm still not seeing any servers?  Which cloudmaster did you use?

**Comment by atavism** [comment]
@oxtoacart I'm looking into it. I used "production-cloudmaster" -- do I need to run ```bin/update.py``` to propagate the new profiles there first?

**Comment by oxtoacart** [comment]
@atavism Weird, it looks like the cloudmaster isn't getting messages (at least not from me).

@aranhoide Any suggestions?  I ran `bin/fake_controller.py launch-fl fl-nl-20150225-001` against the production cloudmaster, and `launch-progress.bash` doesn't show anything being launched.

**Comment by oxtoacart** [comment]
Implemented.  This has yielded an entire order of magnitude improvement in request times for https://www.google.com/humans.txt from within Iran, which bodes very well for other usage as well.  Huge win!

**Comment by myleshorton** [comment]
Awesome!

On Thursday, February 26, 2015, oxtoacart <notifications@github.com> wrote:

> Closed #2210 <https://github.com/getlantern/lantern/issues/2210>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2210#event-242843849>.
>


-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89
